### ISSUE TITLE
DGLHeterograph to list of adjacencies

### ISSUE THREAD

**Issue by bioannidis** [dmlc/dgl#1276]
## 🚀 Feature

Hi. I would like the implementation of the inverse function of 
hetero_from_relations. Given a hetero dgl graph I would like to
obtain a list of adjacency matrices one per relation. I want the 
adjacency matrices to have the same size (even though a lot of
rows and columns might be zero for a specific relation.) The 
current implementation compacts the adjacency matrix and
only contains the nodetypes that participate to a specific 
relation.
## Motivation
The reason I want this feature is to compare my dgl implemented
approach with an existing approach in the literature that uses 
full adjacency matrices.
Vassilis Ioannidis
Applied Scientist Amazon, Palo Alto

**Comment by bioannidis** [comment]
The related ''Bug'' 
+++++++++++++++
follows_g = dgl.graph([(0, 0), (1, 1),(3,3)], 'user', 'follows')
devs_g = dgl.bipartite([(0, 0), (1, 2)], 'developer', 'develops', 'game')
g = dgl.hetero_from_relations([follows_g, devs_g])
adj=g.adjacency_matrix(transpose=True,scipy_fmt='coo',etype='develops')
+++++++++++++++++++++++
here the follows g is 4x4 and the devs_g is 2x2  but the adjacency matrix adj
has shape 2x3, whereas it should be 4x4. This is critical since many implementations assume
that all the nodes are there for each adjacency matrix corresponding to specific relations.

**Comment by BarclayII** [comment]
Hi,
`follows` and `develops` are different edge types, and their source and destination node types are also different, so their adjacency matrices are indeed different in size.
Could you explain why you were expecting the `develops` adjacency matrix to have the same size of `follows`?

**Comment by bioannidis** [comment]
Hi, 
Yes this is what I need. 
Is it possible to obtain this somehow?
Several papers work directly with a full adjacency.
Best,
Vassilis

**Comment by BarclayII** [comment]
You can try converting it to a homogeneous graph and find the edge IDs of one edge type like this:
```python
g = dgl.to_homo(hg)
etype_id = hg.get_etype_id(etype)
eids_of_a_etype = g.filter_edges(lambda edges: edges.data[dgl.ETYPE] == etype_id)
```

**Comment by bioannidis** [comment]
Hi,
Thanks for the help.

I tried this and several variations of you code snippet
and I get different error messages. 
Can you please provide a working example of this code?

Another issue I face is that my DGLHetero graph has different
number of features per node type. So when I call dgl.to_homo(g)
I get the following error message:

{DGLError}Cannot concatenate column features with shape Scheme(shape=(1,), dtype=torch.float32) and shape Scheme(shape=(12,), dtype=torch.float32)

Thank you again.
Best,
Vassilis

**Comment by BarclayII** [comment]
Given the following heterograph:
```python
g = dgl.heterograph({
    ('a', 'ab', 'b'): [(0, 1), (1, 2)],
    ('b', 'ba', 'a'): [(1, 2), (2, 3)]})
```
Confirmed that this works:
```python
g.nodes['a'].data['x'] = torch.randn(4, 5)
g.nodes['b'].data['y'] = torch.randn(3, 4)
g2 = dgl.to_homo(g)
```
But this fails:
```python
g.nodes['a'].data['x'] = torch.randn(4, 5)
g.nodes['b'].data['x'] = torch.randn(3, 4)    # have the same feature name
g2 = dgl.to_homo(g)
```
As a workaround, for now you need to make sure that the features have different names if they are in different sizes.

For DGL improvement, I think a reasonable enhancement is to skip features with the same name but different sizes as well.

**Comment by jermainewang** [comment]
Closed. The feature has been implemented in 0.5.
